### PR TITLE
add ts-scripts bump feature for assets

### PR DESCRIPTION
**Pull request by sotojn**
@busma13 and I added an asset bump feature to the ```ts-scripts``` bump command.

You can now run the bump command inside the ```root``` directory of an ```asset```. 

For example here is running ```yarn bump``` inside of the root directory of ```standard assets```

```cmd
User@standard-assets-root % yarn bump                          
ts-scripts.js bump [packages..]

Update a package to specific version and its dependencies. This should be run in the root of the workspace.

Positionals:
  packages  Run scripts for one or more package (if specifying more than one make sure they are ordered by dependants)
                                                                                                                   [array] [default: []]

Options:
      --prerelease-id  Specify the prerelease identifier, defaults to RC                                                 [default: "rc"]
      --skip-reset     Skip resetting the packages to latest from NPM                                         [boolean] [default: false]
  -d, --deps           Bump the child dependencies recursively, (ignores the monorepo's main package)          [boolean] [default: true]
      --skip-asset     If in an asset repository, bump the package version without updating the asset version [boolean] [default: false]
      --patch          Bump release by patch                                                       [boolean] [required] [default: false]
      --minor          Bump release by minor                                                       [boolean] [required] [default: false]
      --major          Bump release by major                                                       [boolean] [required] [default: false]
      --prerelease     Bump release by prerelease                                                  [boolean] [required] [default: false]
      --prepatch       Bump release by prepatch                                                    [boolean] [required] [default: false]
      --preminor       Bump release by preminor                                                    [boolean] [required] [default: false]
      --premajor       Bump release by premajor                                                    [boolean] [required] [default: false]
  -r, --release        Specify the release change for the version, see https://www.npmjs.com/package/semver.
                       [DEPRECATED] Use --patch, --minor, --major, etc
                                         [string] [choices: "patch", "minor", "major", "prerelease", "prepatch", "preminor", "premajor"]
  -h, --help           Show help                                                                                               [boolean]
  -v, --version        Show version number                                                                                     [boolean]

Examples:
  ts-scripts.js bump  example // 0.20.0 => 0.20.1
  ts-scripts.js bump  --patch example // 0.20.0 => 0.20.1
  ts-scripts.js bump  --minor example // 0.5.0 => 0.6.0
  ts-scripts.js bump  --prepatch example // 0.20.0 => 0.20.1-rc.0
  ts-scripts.js bump  --premajor example // 0.15.0 => 1.0.0-rc.0
  ts-scripts.js bump  --prerelease example // 0.20.1-rc.0 => 0.20.1-rc.1

Error: Bump requires at least one of --patch, --minor, --major, --prerelease, --prepatch, --preminor, --premajor to be specified
    at getRelease (/Users/jsoto/Workspace/teraslice/packages/scripts/dist/src/cmds/bump.js:106:15)
    at Object.handler (/Users/jsoto/Workspace/teraslice/packages/scripts/dist/src/cmds/bump.js:77:25)
    at /Users/jsoto/Workspace/teraslice/packages/scripts/node_modules/yargs/build/index.cjs:1:8993
    at j (/Users/jsoto/Workspace/teraslice/packages/scripts/node_modules/yargs/build/index.cjs:1:4956)
    at _.handleValidationAndGetResult (/Users/jsoto/Workspace/teraslice/packages/scripts/node_modules/yargs/build/index.cjs:1:8962)
    at _.applyMiddlewareAndGetResult (/Users/jsoto/Workspace/teraslice/packages/scripts/node_modules/yargs/build/index.cjs:1:9604)
    at _.runCommand (/Users/jsoto/Workspace/teraslice/packages/scripts/node_modules/yargs/build/index.cjs:1:7231)
    at te.[runYargsParserAndExecuteCommands] (/Users/jsoto/Workspace/teraslice/packages/scripts/node_modules/yargs/build/index.cjs:1:58539)
    at te.parse (/Users/jsoto/Workspace/teraslice/packages/scripts/node_modules/yargs/build/index.cjs:1:40478)
    at te.get [as argv] (/Users/jsoto/Workspace/teraslice/packages/scripts/node_modules/yargs/build/index.cjs:1:62028)
User@standard-assets-root % 
```

In this example I will be adding a patch to ```standard-assets-apis``` package using ```yarn bump standard-assets-apis --patch```

```
User@standard-assets % yarn bump standard-asset-apis  --patch 
⚠  warning   bump has detected the root directory is an Asset.
●  note      bump is in Asset mode.
⚠  warning   @terascope/standard-asset-apis is not in-sync with the remote NPM version, resetting to v0.6.1 before bumping to v0.6.2
ℹ  info      => Updated @terascope/standard-asset-apis from version 0.6.1 to 0.6.2
---> Updating prod dependency @terascope/standard-asset-apis's version of @terascope/standard-asset-apis to 0.6.2
---> Updating dev dependency @terascope/standard-asset-apis's version of @terascope/standard-asset-apis to 0.6.2
ℹ  info      => Updated Asset from version 0.22.1 to 0.22.2
⬤  debug     wrote asset/asset.json JSON file
⬤  debug     wrote asset/package.json JSON file
ℹ  info      => Updated asset.json from version 0.22.1 to 0.22.2
✔  success   

Please commit these changes:

    git commit -a -m "bump: (patch) @terascope/standard-asset-apis@0.6.2" && git push

User@standard-assets % 

````

You can add the ```--skip-asset``` flag to not update the asset version in the process.

### CODE REVIEW

**Comment by sotojn** [comment]
Issue ref #3388

**Review by godber — CHANGES_REQUESTED** [review]

File: packages/scripts/src/helpers/interfaces.ts [hunk 1]
```diff
@@ -116,7 +118,7 @@ export type RootPackageInfo = {
 
 export const AvailablePackageConfigKeys: readonly (keyof PackageConfig)[] = [
     'enableTypedoc', 'testSuite', 'main', 'allowBumpWhenPrivate',
-    'linkToMain', 'root', 'tests'
+    'linkToMain', 'root', 'tests', 'asset'
```
[the remark this answers is not in the document]
Comment (busma13) [reply packages/scripts/src/helpers/interfaces.ts:121]: No. Removed.

File: packages/scripts/src/helpers/packages.ts [hunk 2]
```diff
@@ -58,19 +58,22 @@ export function listPackages(
...
+    let workspaces = (
         Array.isArray(rootPkg.workspaces)
             ? rootPkg.workspaces
             : rootPkg.workspaces.packages
     ).slice();
 
     if (!workspaces) return [];
 
+    ///    Testing in the case where one of the workspaces is the root directory
+    workspaces = workspaces.filter((space) => space !== '.');
```
Comment (godber) [inline packages/scripts/src/helpers/packages.ts:71]: Question: Should we be doing this, or should we be be removing the `.` from the workspaces of the assets?  Why does this discrepancy exist?

File: packages/scripts/test/bump-spec.ts [hunk 3]
```diff
@@ -370,3 +372,349 @@ describe('Bump Utils', () => {
...
+                deps: true,
+                skipReset: true,
+                packages: [cloneDeep(pkg1!)]
+            };
+            let result: Record<string, BumpPkgInfo>;
+
+            beforeAll(async () => {
+                result = await getPackagesToBump(testPackages, options);
+                /// Create an asset folder with asset.json
+                const path = `${process.cwd()}/asset`;
```
Comment (godber) [inline packages/scripts/test/bump-spec.ts:442]: Can this be changed to use a temp directory?

Comment (godber) [reply packages/scripts/test/bump-spec.ts:442]: https://nodejs.org/docs/latest-v18.x/api/fs.html#fspromisesmkdtempprefix-options

File: docs/packages/test-pkg/overview.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+---
```
[the remark this answers is not in the document]
Comment (busma13) [reply docs/packages/test-pkg/overview.md]: Removed.

File: packages/scripts/src/helpers/bump/index.ts [hunk 5]
```diff
@@ -42,3 +45,72 @@ Please commit these changes:
     git commit -a -m "${commitMsgs.join('" -m "')}" && git push
 `);
 }
+
```
[the remark this answers is not in the document]
Comment (busma13) [reply packages/scripts/src/helpers/bump/index.ts:55]: `bumpPackagesForAsset` was removed and the functionality was added into `bumpPackages`.

File: packages/scripts/test/bump-spec.ts (hunk 3 above)
Comment (busma13) [reply packages/scripts/test/bump-spec.ts]: Done.

File: packages/scripts/src/helpers/packages.ts (hunk 2 above)
Comment (busma13) [reply packages/scripts/src/helpers/packages.ts:70]: After talking with Jared, we will need to make the test runner asset aware before we can remove the `.` from workspaces.

**Review by godber — CHANGES_REQUESTED** [review]
Great work!

After some manual review we decided to consider the following changes:

* skip doc generation for assets
* skip initial munging of package.json and tsconfig for assets

I also observed an error when running bump-asset, the top level package.json did not get it's version bumped, only the `asset/package.json` and `asset/asset.json`.

File: packages/scripts/src/cmds/bump-asset.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        const release = getRelease(argv);
+        const rootInfo = getRootInfo();
+
+        await syncAll({
+            verify: true,
+            tsconfigOnly: rootInfo.terascope.version === 2,
+            isAsset: true
+        });
+
+        if (!rootInfo.terascope.asset) {
```
Comment (godber) [inline packages/scripts/src/cmds/bump-asset.ts:51]: Move above `syncAll`

**Review by godber — CHANGES_REQUESTED** [review]

**Review by godber — APPROVED** [review]
This looks great, thanks guys

File: packages/scripts/src/cmds/bump-asset.ts [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import { getRootInfo } from '../helpers/misc';
+import signale from '../helpers/signale';
+
+const releaseChoices: readonly ReleaseType[] = [
+    'patch', 'minor', 'major', 'prerelease', 'prepatch', 'preminor', 'premajor'
+];
+
+const cmd: CommandModule = {
+    command: 'bump-asset',
+    describe: 'Update an asset to specific version. This should be run in the root of the workspace.',
```
Comment (jsnoble) [inline packages/scripts/src/cmds/bump-asset.ts:14]: The describe statement is a bit confusing. This seems to describe the exact same thing as the bump command which will lead to confusion. This statement will show up when using the --help command so this needs to better describe how this is different from bump and when this should be used over bump

File: packages/scripts/src/cmds/bump.ts [hunk 8]
```diff
@@ -36,6 +37,11 @@ const cmd: CommandModule = {
                 description: "Bump the child dependencies recursively, (ignores the monorepo's main package)",
                 default: true,
                 type: 'boolean',
+            })
+            .option('skip-asset', {
+                description: 'If in an asset repository, bump the package version without updating the asset version',
```
Comment (jsnoble) [inline packages/scripts/src/cmds/bump.ts:42]: Which package version does it do?  The top level package.json version usually/should mirror the asset version and should be in sync

File: packages/scripts/src/cmds/bump.ts [hunk 9]
```diff
@@ -74,20 +80,31 @@ const cmd: CommandModule = {
...
-        await syncAll({ verify: true, tsconfigOnly: rootInfo.terascope.version === 2 });
+
+        await syncAll({
+            verify: true,
+            tsconfigOnly: rootInfo.terascope.version === 2,
+            isAsset: rootInfo.terascope.asset
+        });
+
+        if (rootInfo.terascope.asset) {
+            signale.warn('bump has detected the root directory is an Asset.');
```
Comment (jsnoble) [inline packages/scripts/src/cmds/bump.ts:91]: is there a reason on having two different logging statements here? The first one should not be a warning, it should be at the info level or debug level of logging. Warn should be used if something is off or potentially could be off or run in a dangerous manner

File: packages/scripts/src/helpers/misc.ts [hunk 10]
```diff
@@ -39,41 +39,65 @@ export function getRootDir(cwd: string = process.cwd()): string {
...
-                suites: {}
+    if (isRootAsset) {
+        return sortPackageJson(defaultsDeep(pkg, {
+            dir,
+            relativeDir: '.',
+            folderName,
+            displayName: getName(pkg.name),
+            engines: {
+                node: '>=14.17.0',
+                yarn: '>=1.16.0'
```
Comment (jsnoble) [inline packages/scripts/src/helpers/misc.ts:57]: since we are changing things, the minimum of yarn if not set should be ` "yarn": ">=1.22.19"`.  We have not updated these things as we should. We have libraries that fail if its less than that

**Review by jsnoble — CHANGES_REQUESTED** [review]

File: packages/scripts/src/cmds/bump.ts (hunk 9 above)
Comment (busma13) [reply packages/scripts/src/cmds/bump.ts:91]: I've changed this to a single info statement.

File: packages/scripts/src/helpers/misc.ts (hunk 10 above)
Comment (busma13) [reply packages/scripts/src/helpers/misc.ts:57]: I will also update on line 81.

File: packages/scripts/src/cmds/bump.ts (hunk 8 above)
Comment (busma13) [reply packages/scripts/src/cmds/bump.ts:42]: I think the trouble is differentiating between packages in the `/packages` directory and package.json files in general. Does this clear it up?
```
description: 'If in an asset repository, bump the version of the specified package(s), but do not update the asset version in \'./package.json\', \'./asset/asset.json\', and \'./asset/package.json\'.',
```

File: packages/scripts/src/cmds/bump-asset.ts [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+            .option('prerelease-id', {
+                default: 'rc',
+                description: 'Specify the prerelease identifier, defaults to RC',
+            });
+
+        releaseChoices.forEach((choice, i, arr) => {
+            const otherChoices = arr.filter((_item, index) => index !== i);
+
+            y = y.option(choice, {
+                description: `Bump release by ${choice}`,
```
[the remark this answers is not in the document]
Comment (busma13) [reply packages/scripts/src/cmds/bump-asset.ts:31]: I just copied that straight from bump.ts. Something more like this maybe?
```
      --patch          Bump asset to next patch version 
      --minor          Bump asset to next minor version
      --major          Bump asset to next major version
      --prerelease     Bump asset to next prerelease version
      --prepatch       Bump asset to next prepatch version
      --preminor       Bump asset to next preminor version 
      --premajor       Bump asset to next premajor version
```

File: packages/scripts/src/cmds/bump.ts (hunk 8 above)
Comment (godber) [reply packages/scripts/src/cmds/bump.ts:42]: Start out more concretely, not with an `If`:

```
Bumps a package version and all of the places that package is used in the current repository.
...
```

File: packages/scripts/src/cmds/bump-asset.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (godber) [reply packages/scripts/src/cmds/bump-asset.ts:31]: That is better.

File: packages/scripts/src/cmds/bump-asset.ts (hunk 7 above)
Comment (busma13) [reply packages/scripts/src/cmds/bump-asset.ts:14]: This is what Austin and I came up with. It might be a bit wordy...
```
Update only the asset version.
This is for changes in the asset code. Changes in packages should use the bump command.
This will update the version number in './package.json', './asset/asset.json', and './asset/package.json'.
This should be run in the root of the workspace.
```

File: packages/scripts/src/cmds/bump.ts (hunk 8 above)
Comment (busma13) [reply packages/scripts/src/cmds/bump.ts:42]: Does this look better?
```
      --skip-asset     Bumps a package version and all of the places that package is used in the current repository.
                       If in an asset repository, do not update the asset version in './package.json', './asset/asset.json', and './asset/package.json'.
```

Comment (godber) [reply packages/scripts/src/cmds/bump.ts:42]: Oh, I was confused, I thought that was the description for `bump` ... I hadn't realized it was for `skip_asset`, so yours is better, just drop the details.

```
description: 'If in an asset repository, bump the version of the specified package(s), but do not update the asset version.',
```

Comment (godber) [reply packages/scripts/src/cmds/bump.ts:42]: Actually, wait, that's not all it does.

**Review by godber — APPROVED** [review]
